pkg/operator/noobaa/nbapi/rpc.go

```diff
@@ -0,0 +1,151 @@
+*/
+
+package nbapi
+
+import (
+	"bytes"
+	"crypto/tls"
+	"encoding/json"
+	"io/ioutil"
+	"net/http"
+	"time"
+)
+
+// RPCClient makes API calls to noobaa.
+// Requests to noobaa are plain http requests with json request and json response.
+type RPCClient struct {
+	Router     RPCRouter
+	HTTPClient http.Client
+	AuthToken  string
+}
+
+// RPCRequest is the interface for request structs.
+// Structs that implement it will build a request body and a response to be decoded.
+type RPCRequest interface {
+	build() (*RPCRequestBody, RPCResponse)
+}
+
+// RPCResponse is the interface for response structs.
+// RPCResponseBody is the only real implementor of it.
+// Specific response structures should just include RPCResponseBody inline.
+type RPCResponse interface {
+	GetError() *RPCError
+}
+
+// RPCRequestBody is the structure encoded in every request
+// Params is the specific request params.
+type RPCRequestBody struct {
+	API       string      `json:"api,omitempty"`
+	Method    string      `json:"method,omitempty"`
+	AuthToken string      `json:"auth_token,omitempty"`
+	Params    interface{} `json:"params,omitempty"`
+}
+
+// RPCResponseBody is the structure encoded in every response
+// The specific response structure should include this inline,
+// and add the standard Reply field with the specific fields.
+type RPCResponseBody struct {
+	Op        string    `json:"op,omitempty"`
+	RequestID string    `json:"reqid,omitempty"`
+	Took      float64   `json:"took,omitempty"`
+	Error     *RPCError `json:"error,omitempty"`
+}
+
+// GetError is implementing RPCResponse interface
+func (r *RPCResponseBody) GetError() *RPCError { return r.Error }
+
+// RPCError is a struct sent by noobaa servers to denote an error response.
+type RPCError struct {
+	RPCCode string `json:"rpc_code"`
+	Message string `json:"message"`
+}
+
+// Error is implementing the standard error type interface
+func (e *RPCError) Error() string { return e.Message }
+
+var _ error = &RPCError{}
+
+// NewClient initializes an RPCClient with defaults
+func NewClient(router RPCRouter) *RPCClient {
+	return &RPCClient{
+		Router: router,
+		HTTPClient: http.Client{
+			Timeout: 120 * time.Second,
+			Transport: &http.Transport{
+				TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
+			},
+		},
+	}
+}
+
+// Call an API method to noobaa.
+// The response type should be defined to include RPCResponse inline - see below the specific response types.
+// This is needed in order for json.Unmarshal() to decode into the reply structure.
+func (c *RPCClient) Call(req RPCRequest) error {
+
+	body, res := req.build()
+	api := body.API
+	method := body.Method
+	if body.AuthToken == "" {
+		body.AuthToken = c.AuthToken
+	}
+	address := c.Router.GetAddress(api)
+	logger.Infof("NBAPI: %s.%s - RPC Call to %s: %+v", api, method, address, req)
+
+	reqBytes, err := json.Marshal(body)
+	if err != nil {
+		logger.Errorf("NBAPI: ❌ %s.%s - Encoding request failed: %s", api, method, err)
```

Comment:
are these emoji chars in the log error strings?  i'm not sure if that's part of the string or if github is rendering this diff weird haha

does having non ascii chars in log strings potentially break any logging systems that may be collecting/aggregating/viewing these logs?

Reply:
Yes, these are unicode characters to make the logs more human readable.
I certainly hope it will not break machine readability because unicode chars might be used in resource names given by users too, right?
So I think to keep it for now and wait for first pushback.
Thanks